src/Dhall/Map.hs

```diff
@@ -0,0 +1,258 @@
+{-# LANGUAGE CPP                #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveTraversable  #-}
+{-# LANGUAGE RecordWildCards    #-}
+
+-- | `Map` type used to represent records and unions
+
+module Dhall.Map
+    ( -- * Type
+      Map
+
+      -- * Construction
+    , singleton
+    , fromList
+
+      -- * Sorting
+    , sort
+    , isSorted
+
+      -- * Insertion
+    , insert
+    , insertWith
+
+      -- * Deletion/Update
+    , delete
+
+      -- * Query
+    , lookup
+    , member
+
+      -- * Combine
+    , union
+    , unionWith
+    , intersection
+    , intersectionWith
+
+      -- * Traversals
+    , mapWithKey
+    , traverseWithKey
+    , foldMapWithKey
+
+      -- * Conversions
+    , toList
+    , toMap
+    , keys
+    ) where
+
+import Control.Applicative ((<|>))
+import Data.Data (Data)
+import Data.Semigroup
+import Prelude hiding (lookup)
+
+import qualified Data.Map
+import qualified Data.Set
+
+{-| A `Map` that remembers the original ordering of keys
+
+    This is primarily used so that formatting preserves field order
+
+    This is done primarily to avoid a dependency on @insert-ordered-containers@
+    and also to improve performance
+-}
+data Map k v = Map (Data.Map.Map k v) [k]
+    deriving (Data, Eq, Foldable, Functor, Show, Traversable)
```

Comment:
Do we use the `Eq` instance? That's semi-dubious for this type, since we can reasonably want either strict equality of key order, or equality up to key order.

Reply:
It's used by `Expr` needing an `Eq` instance, since `Record` and `Union` are `Map`s under the hood

Reply:
I'm unsure that I follow. Would `Eq` not check that the list is in order as well?

Reply:
Have a look at the [definition of the instances for `InsOrdHashMap`](http://hackage.haskell.org/package/insert-ordered-containers-0.2.1.0/docs/src/Data-HashMap-Strict-InsOrd.html#InsOrdHashMap): it totally tosses away the indices of the elements. That's the code Dhall's currently using, meaning `Record` and `Union` compare for equality up to key insertion order, but this PR is going to change that.

Reply:
As far as I can see it only compares values inserted into the underlying `HashMap`. Maybe I'm misunderstanding something? :/

Reply:
Yes, that's what the `InsOrdHashMap` instance does. However, the derived `Dhall.Map.Map` instance is going to check for key ordering as well. To match the previous behaviour, we'd need to write something like:

```
instance (Eq k, Eq v) => Eq (Map k v) where
  Map a _ == Map b _ = a == b
```

Reply:
Right right, I've done that here actually :smile: https://github.com/FintanH/dhall-haskell/blob/fintan/new-map/src/Dhall/Map.hs#L55-L57

Reply:
@FintanH @quasicomputational: I prefer to use the new order-sensitive `Eq` instance.  In other words, I view a `Dhall.Map.Map` as an ordered association list that just happens to provide better time complexity for element lookup.

The main reason why is that this is more consistent with the standard, which no longer requires that record/union equality is order insensitive (See: https://github.com/dhall-lang/dhall-lang/pull/223).  Instead, whenever order-insensitivity is required the standard requires an explicit β-normalization step which (among other things) sorts the records/unions.